crates/project-model/src/tests.rs

```diff
@@ -1666,3 +208,36 @@ fn rust_project_is_proc_macro_has_proc_macro_dep() {
     // on the proc_macro sysroot crate.
     crate_data.dependencies.iter().find(|&dep| dep.name.deref() == "proc_macro").unwrap();
 }
+
+#[test]
+fn crate_graph_dedup_identical() {
+    let (mut crate_graph, proc_macros) =
+        load_cargo_with_sysroot(&mut Default::default(), "regex-metadata.json");
+    crate_graph.sort_deps();
+
+    let (d_crate_graph, mut d_proc_macros) = (crate_graph.clone(), proc_macros.clone());
+
+    crate_graph.extend(d_crate_graph.clone(), &mut d_proc_macros);
+    assert!(crate_graph.iter().eq(d_crate_graph.iter()));
+    assert_eq!(proc_macros, d_proc_macros);
+    check_crate_graph(
+        crate_graph,
+        expect_file!["../test_data/output/crate_graph_dedup_identical.txt"],
+    );
+}
+
+#[test]
+fn crate_graph_dedup() {
+    let path_map = &mut Default::default();
+    let (mut crate_graph, _proc_macros) =
+        load_cargo_with_sysroot(path_map, "ripgrep-metadata.json");
+    assert_eq!(crate_graph.iter().count(), 81);
+    crate_graph.sort_deps();
+    let (regex_crate_graph, mut regex_proc_macros) =
+        load_cargo_with_sysroot(path_map, "regex-metadata.json");
+    assert_eq!(regex_crate_graph.iter().count(), 60);
+
+    crate_graph.extend(regex_crate_graph, &mut regex_proc_macros);
+    assert_eq!(crate_graph.iter().count(), 118);
+    check_crate_graph(crate_graph, expect_file!["../test_data/output/crate_graph_dedup.txt"])
+}
```

Comment:
So no one has to sift through the massive diff due to test files, this is where the tests are :)